app/controllers/shipments_controller.rb

```diff
@@ -0,0 +1,68 @@
+require "timeout"
+class ShipmentsController < ApplicationController
+
+  skip_before_filter :verify_authenticity_token, only: :ship
+
+  def index
+    render json: { errors: "This page does not exist. Perhaps you were looking for 'adaships.herokuapp.com/log' or 'adaships.herokuapp.com/ship'."}, status: 404
+  end
+
+  def ship
+    @shipping_data = JSON.parse(request.body.read)
+
+    if  @shipping_data.keys.length != 5 ||
+        @shipping_data.values.select(&:nil?).length != 0 ||
+        @shipping_data.values.select(&:empty?).any?
+      render json: { errors: "Incomplete request."}, status: :bad_request
+    else
+      ups_response = response_data(ups_login)
+      usps_response = response_data(usps_login, true)
+
+      if ups_response.empty? || usps_response.empty?
+        render json: { errors: "Request timed out."}, status: 408
+      else
+        response = { ups: ups_response, usps: usps_response }
+        render json: response.as_json
+      end
+    end
+  end
+
+  private
+
+  def response_data(login_method, usps = false)
+    begin
+      response = Timeout::timeout(TIMEOUT) {
+        login_method.find_rates(origin, destination, package(@shipping_data[:weight]))
+      }
+
+      if usps
+        response = response.rates.select { |rate|
+          rate.service_code == "1" ||
+          rate.service_code == "3" ||
+          rate.service_code == "4" ||
+          rate.service_code == "22"
+        }
```

Comment:
This is the same logic that is in the `usps_shipments_controller`. Can you encapsulate it somewhere?